### PR TITLE
ZJIT: Support guarding *Exact types

### PR DESCRIPTION
**Pull request by st0012**
ZJIT already can generate guard type instructions for *Exact types.

For example:

```
def test(strings)
  strings.map do |string|
    string.bytesize
  end
end

test(["foo", "bar"])
```

```
HIR:
fn block in test:
bb0(v0:BasicObject, v1:BasicObject):
  PatchPoint MethodRedefined(String@0x1014be890, bytesize@0x19f1)
  v7:StringExact = GuardType v1, StringExact
  v8:Fixnum = CCall bytesize@0x16fa4cc18, v7
  Return v8

```

But zjit only supported guarding fixnums so this script would panic.

This commit adds support for guarding *Exact types.

### CODE REVIEW

File: zjit/src/codegen.rs [hunk 1]
```diff
@@ -979,6 +979,16 @@ fn gen_guard_type(jit: &mut JITState, asm: &mut Assembler, val: lir::Opnd, guard
         // Check if opnd is Fixnum
         asm.test(val, Opnd::UImm(RUBY_FIXNUM_FLAG as u64));
         asm.jz(side_exit(jit, state)?);
+    } else if let Some(expected_class) = guard_type.runtime_exact_ruby_class() {
+        // Guard for exact class types (e.g., ClassExact, ArrayExact, StringExact, etc.)
+        asm_comment!(asm, "guard exact class");
+
+        // Get the class of the value
+        let klass = asm.ccall(rb_yarv_class_of as *const u8, vec![val]);
```
Comment (tekknolagi) [inline zjit/src/codegen.rs:981]: If we can inline this somehow, that would be great (needs a check if it's a special const, then load `RBASIC(obj)->klass`)

File: zjit/src/codegen.rs [hunk 2]
```diff
@@ -979,6 +979,16 @@ fn gen_guard_type(jit: &mut JITState, asm: &mut Assembler, val: lir::Opnd, guard
...
+
+        // Compare with expected class
+        asm.cmp(klass, Opnd::UImm(expected_class.into()));
```
Comment (tekknolagi) [inline zjit/src/codegen.rs]: `Value`, not `UImm`

File: test/ruby/test_zjit.rb [hunk 3]
```diff
@@ -915,6 +915,51 @@ def test_instruction_order
...
+    }, call_threshold: 2
+  end
+
+  def test_guard_type_side_exit
+    # This test demonstrates that the guard side exit works correctly
+    # In this case, when we call with a non-Class object, it should fall back to interpreter
+    assert_compiles '["String", "Integer", "Foo"]', %q{
+      class Foo
+        def name
+          "Foo"
```
Comment (tekknolagi) [inline test/ruby/test_zjit.rb]: Let's use a different value than the class name to disambiguate

File: zjit/src/codegen.rs (hunk 1 above)
Comment (tekknolagi) [reply zjit/src/codegen.rs:981]: (this is a very handwavy sketch)

```
if expected_class == FalseClassExact {
  cmp val, 0
  jnz interpreter
} else if expected_class == NilClassExact {
  // fast check for nil
}
...
else {
  assert!(!expected_class.special_const_p());
  test val, RUBY_IMMEDIATE_MASK
  jz interpreter
  load cls, val->klass
  cmp cls, expected_class
  jnz interpreter
}
```

Comment (tekknolagi) [reply zjit/src/codegen.rs:981]: Could be a TODO

File: zjit/src/hir.rs [hunk 4]
```diff
@@ -6517,4 +6517,44 @@ mod opt_tests {
...
+            bb0(v0:BasicObject, v1:BasicObject):
+              PatchPoint MethodRedefined(Class@0x1000, name@0x1008)
+              v7:ClassExact = GuardType v1, ClassExact
+              v8:StringExact|NilClassExact = CCall name@0x1010, v7
+              Return v8
+        "#]]);
+    }
+
+    #[test]
+    fn test_guard_string_for_bytesize() {
```
Comment (tekknolagi) [inline zjit/src/hir.rs]: can you add tests for immediates like false, nil, true, symbol, float

**Review by tekknolagi — APPROVED** [review]
lgtm

File: zjit/src/hir.rs (hunk 4 above)
Comment (st0012) [reply zjit/src/hir.rs]: I added them in https://github.com/ruby/ruby/pull/13825

Comment (tekknolagi) [reply zjit/src/hir.rs]: for posterity, per my comment there, you will also need to update this PR to actually codegen and execute the optimized code

File: zjit/src/codegen.rs (hunk 1 above)
Comment (st0012) [reply zjit/src/codegen.rs:981]: I've added Ruby execution tests for different classes (integer, float, true, false...etc.) but for inlining I plan to add them in a separate PR. I created https://github.com/Shopify/ruby/issues/624 for it.
